packages/core/src/derived/processor.ts

```diff
@@ -0,0 +1,237 @@
+    oldData: DerivedData[],
+    newData: DerivedData[],
+    storage: Storage & TxOperations
+  ): Promise<void> {
+    // Do diff from old refs to remove only missing.
+    const { additions, updates, deletes } = findExistingData(oldData, newData)
+
+    const added = Promise.all(additions.map(async (add) => await storage.createDoc(add._class, add.space, add)))
+    const updated = Promise.all(updates.map(async (up) => await storage.updateDoc(up._class, up.space, up._id, up)))
+    const deleted = Promise.all(deletes.map(async (del) => await storage.removeDoc(del._class, del.space, del._id)))
+
+    await Promise.all([added, updated, deleted])
+  }
+
+  private async applyMapper (descriptor: Descr, tx: Tx): Promise<DerivedData[] | undefined> {
+    const mapper = derivedDataMappers.get(descriptor.mapper ?? EMPTY_MAPPER)
+    if (mapper !== undefined) {
+      return await mapper.map(tx, { descriptor, hierarchy: this.hierarchy, storage: this.storage })
+    }
+    return undefined
+  }
+
+  private async applyRules (
+    d: Descr,
+    doc: Doc,
+    ruleFields: { value: string, rule: MappingRule }[]
+  ): Promise<DerivedData[]> {
+    const results: DerivedData[] = []
+    // Check if we have rules matched.
+
+    for (const { rule, value } of ruleFields) {
+      if (rule.pattern !== undefined) {
+        this.processRulePattern(d, doc, rule.pattern, rule.targetField, value, results)
+      } else {
+        // Just copy of value
+        this.processAssignValue(results, d, doc, rule.targetField, value)
+      }
+    }
+    return results
+  }
+
+  private processAssignValue (results: DerivedData[], d: Descr, doc: Doc, targetField: string, value: string): void {
+    const lastResult = lastOrAdd(results, d, doc)
+    ;(lastResult as any)[targetField] = value
+  }
+
+  private processRulePattern (
+    d: Descr,
+    doc: Doc,
+    pattern: RuleExpresson,
+    targetField: string,
+    sourceValue: string,
+    results: DerivedData[]
+  ): DerivedData[] {
+    // Extract some data from value
+    const reg = new RegExp(pattern.pattern, 'g')
+
+    let matches: RegExpExecArray | null
+    let needAdd = false
+    while ((matches = reg.exec(sourceValue)) !== null) {
+      if (needAdd) {
+        // We have multi doc, so produce document and wait for next match.
+        results.push(newDerivedData(lastOrAdd(results, d, doc), d))
+        needAdd = false
+      }
+      const lastResult = lastOrAdd(results, d, doc)
+      ;(lastResult as any)[targetField] = groupOrValue(pattern, matches)
+
+      if (pattern.multDoc ?? false) {
+        needAdd = true
+      }
+    }
+    return results
+  }
+
+  private async refreshDerivedData (d: Descr, modifiedBy: Ref<Account>): Promise<void> {
+    // Perform a full rebuild of derived data of required type.
+    const ops = withOperations(modifiedBy, this.storage)
+
+    // we need to find all objects affected
+
+    const allDD = await this.storage.findAll(d.sourceClass, {})
```

Comment:
As discussed on today call, we need API for partial data processing, since it could be a lot of data. #165 could be place to discuss it.